_posts/2025-03-18-movie-similarity-search-using-vector-databases.adoc

```diff
@@ -0,0 +1,416 @@
+to avoid adding irrelevant results to the response.
+
+[source,java]
+----
+package io.quarkiverse.langchain4j.sample;
+
+import java.util.List;
+
+import dev.langchain4j.data.embedding.Embedding;
+import dev.langchain4j.data.segment.TextSegment;
+import dev.langchain4j.model.embedding.EmbeddingModel;
+import dev.langchain4j.store.embedding.EmbeddingSearchRequest;
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.transaction.Transactional;
+
+@ApplicationScoped
+public class MovieRecommendationService {
+
+  @Inject
+  EmbeddingStore<TextSegment> embeddingStore;
+
+  @Inject 
+  EmbeddingModel embeddingModel;
+
+  @Transactional
+  public List<Movie> searchSimilarMovies(String overview) {
+    
+    Embedding embedding = embeddingModel.embed(overview).content();
+    EmbeddingSearchRequest request = EmbeddingSearchRequest.builder()
+    .queryEmbedding(embedding)
+    .minScore(0.5)
+    .maxResults(10)
+    .build();
+
+    return embeddingStore.search(request).matches().stream().map(m -> {
+      Long id = m.embedded().metadata().getLong("id");
+      Movie movie = Movie.findById(id);
+      return movie;
+    }).toList();
+
+  }
+}
+----
+
+
+==== Loading the movies ====
+To populate the movies in the database, we'll use a CSV file containing the top 1000 movies from IMDB.
+The important columns for us are:
+
+* `title` the movie title
+* `overview` the movie overview
+* `link` the link to the image poster
+
+So, a simplified version of our Movie entity object looks like: 
+
+[source,java]
+----
+package io.quarkiverse.langchain4j.sample;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+import jakarta.persistence.Entity;
+import java.util.List;
+
+@Entity
+public class Movie extends PanacheEntity {
+
+    public String link;
+    public String title;
+    public String overview;
+
+    public static List<Movie> searchByTitleLike(String title) {
+      return find("title like ?1", "%" + title + "%").list();
+    }
+}
+----
+
+To actually parse the CSV we'll need a utility that reads a line and extracts title, overvie and link.
+
+One thing to note is that the CSV is delimited by comas, but commas may also be present in the overview, title etc
+enclosed in double quotes. 
+
+Another thing to note is that in some cases the overview may be too long. For simplicity we'll truncate it to 255 characters.
+
+[source,java]
+----
+    public static Movie fromCsvLine(String line) {
+       String values [] = line.split(",(?=([^\"]*\"[^\"]*\")*[^\"]*$)");
+       Movie movie = new Movie();
+       movie.link = values[0].replace("\"", "");
+       movie.title = values[1].replace("\"", "");
+       movie.overview = values[7].replace("\"", "");
+       if (movie.overview.length() > 255) {
+           movie.overview = movie.overview.substring(0, 255);
+       }
+       return movie;
+    }
```

Comment:
Perhaps we can use something like we do at https://github.com/quarkiverse/quarkus-langchain4j/blob/main/samples/sql-chatbot/src/main/java/io/quarkiverse/langchain4j/sample/chatbot/CsvIngestorExample.java?